### PR TITLE
Use modal kinds to do mode crossing

### PR DESCRIPTION
**Pull request by goldfirere**
As title.

This is written as a series of commits. I recommend reviewing the commits in order.

I believe the testsuite fully passes after each commit.

Some later commits remove uses of `is_immediate`. Some of the changes there removed checks that the testsuite didn't notice -- and so I assumed are redundant. Careful scrutiny invited.

I'm a little worried about the performance impact here, given the uses of the expensive `type_jkind_purely`. But I think we've already paid the cost, because the old check would have been just about as bad. (The one scenario where the new check is worse is when you have a `[@@immediate]` abstract type that keeps expanding further. Previously, this short-circuited. Now, it won't.) It would be possible to improve this back to the old performance characteristics by stopping when a modal upper bound hits the minimum. Other work I have planned will make that easier to do later, though.

Recommended review: @riaqn

### CODE REVIEW

File: ocaml/typing/mode_intf.mli
Comment (goldfirere) [reply ocaml/typing/mode_intf.mli:29]: I agree that `eq` and `le` (together) are equivalent in expressiveness to `less_or_equal`. An alternative implementation could make a `Minimal_lattice` module type without `equal` and `le`, a `Lattice` module type with them, and a functor `Make_lattice (L : Minimal_lattice) : Lattice with type t = L.t` that derives them. Because they're so simple, though, I thought this extra structure wasn't worth it. Or we could drop `equal` and `le` and just use `less_or_equal` everywhere, but that interface is harder to work with.

In terms of usefulness: the kind infrastructure wants to know the difference between "less than" and "equal", because it helps guide error messages. Specifically, when we compute the meet between `m1` and `m2`, if they're equal, the operation is uninteresting (and should not change error messages). If, on the other hand, the new mode is really less than the old one, we've learned something materially new, and we want to update error messages accordingly.

For `mode_cross_left`, we could actually just use `le` -- that one is not a problem.

Comment (goldfirere) [reply ocaml/typing/mode_intf.mli:29]: For kinds, I care only about constants. I don't think we'll need mode variables in kinds.

And, yes, it's true that `le` is sufficient, as you say. But I think it's more efficient to implement `le` in terms of `less_or_equal` than the other way around. In particular, I think if we inline `less_or_equal` (and `Le_result.is_le`) into `le`, we'll get the same code as now, but I don't see how the optimizer could do something similar with two calls to `le`.

Comment (riaqn) [reply ocaml/typing/mode_intf.mli:29]: >For kinds, I care only about constants. I don't think we'll need mode variables in kinds.

Maybe I misunderstood - I thought in your original message, you meant that the kind system wants to know if mode crossing `m_new = meet m b`indeed has changed the mode `m` of some value. In that case, the `m` could often be a mode variable.

I believe `Le_result` is used in both the kind and mode system. I'm not sure about kinds. But I think for modes, `le` alone is mostly enough, because:
1. Most of time, we really just want to know if `a <= b`, instead of `a = b`.
2. In the rare cases where we do want to know `a = b`, highly likely that we already know `a <= b`, and it's sufficient to just test `b <= a` to know `a = b`.

In both cases, it's more efficient to run `le` directly instead of going through `less_or_equal`.

Going back to this particular PR, I believe `Mode.equal` is used only for `mode_cross_left` and `expect_mode_cross`, both can be adapted to use `Mode.le` instead. For example:
```
Uniqueness.Const.equal upper_bounds.uniqueness Uniqueness.Const.min
```
can be written as 
```
Uniqueness.Const.le upper_bounds.uniqueness Uniqueness.Const.min
```

This is how `solver.ml` test something for min/max and optimise accordingly.

File: ocaml/typing/ctype.ml
Comment (riaqn) [inline ocaml/typing/ctype.ml:2039]: missing `]` here.

Also would it be better to factor this type to `(unit, type_jkind_sub_error) Result.t`?

File: ocaml/typing/ctype.ml
Comment (riaqn) [inline ocaml/typing/ctype.ml:2046]: nit: the `fuel` argument is no longer visible externally - might move this comment to the `loop` below.

File: ocaml/testsuite/tests/typing-local/local.ml
Comment (riaqn) [inline ocaml/testsuite/tests/typing-local/local.ml:1823]: something is wrong here? I suspect this is because the changes to `Tpat_constant`

File: ocaml/typing/ctype.ml
Comment (riaqn) [inline ocaml/typing/ctype.ml:5704]: Any chance we might benefit from a principality check here?

File: ocaml/testsuite/tests/typing-modal-kinds/basics.ml
Comment (riaqn) [inline ocaml/testsuite/tests/typing-modal-kinds/basics.ml:273]: Thanks for the comment - I suppose the scope of this PR is only about immediate, and therefore "arrow types crosses linearity" is left as future PRs? I'm fine with that - just confirming (because the title of this PR suggests otherwise)

File: ocaml/testsuite/tests/typing-modal-kinds/basics.ml
Comment (riaqn) [inline ocaml/testsuite/tests/typing-modal-kinds/basics.ml:146]: Should unboxed floats crosses all three axes? Just confirming that we are leaving this as future work.

Similarly, this tests contains many tests that aren't related to immediacy checks, those tests will only be relevant for future PRs, right?

File: ocaml/typing/ctype.ml
Comment (riaqn) [reply ocaml/typing/ctype.ml:5502]: I think it might not be testable, because this is used by `enlarge_type`, which is used in `typecore.ml` but falls back to `subtype` which still crosses mode.

File: ocaml/typing/mode_intf.mli
Comment (goldfirere) [reply ocaml/typing/mode_intf.mli:29]: > I'm not sure about kinds.

It's true that `le` is sufficient for modes. But it's not sufficient for kinds.

`Jkind.combine_histories` uses `Desc.sub` (and cares whether `Desc.sub` returns `Less` or `Equal`). `Desc.sub` calls `Const.sub`, which calls `Mode.Alloc.Const.less_or_equal` -- still caring about the difference between `Less` and `Equal`. That's why I added the new functionality.

File: ocaml/typing/mode.ml
Comment (goldfirere) [reply ocaml/typing/mode.ml:97]: When/if we add a third mode, it will no longer be redundant. I'm happy to explain why I've kept it, but keeping was intentional. I don't feel strongly here.

File: ocaml/typing/jkind.mli
Comment (goldfirere) [reply ocaml/typing/jkind.mli:40]: I actually did this already. See https://github.com/goldfirere/flambda-backend/commit/d802597fbdaaa850e1ed9209a1305c5dcdf71e17  But there is nothing to make good use of the work... so it seemed a bit silly. Someday, we'll actually use externality in libraries, but not for a while. So I thought it was fine to keep `Externality` in jkind, though I suppose we could just move it to Mode, like you say. I favor just leaving it alone, but not strongly.

File: ocaml/typing/ctype.ml
Comment (goldfirere) [reply ocaml/typing/ctype.ml:5502]: I deleted it because it called `is_immediate`.... but the property of being "immediate" doesn't really exist any more, and I didn't immediately see how to translate this into the new world of modal kinds. So I deleted it to see what broke so that I might understand it better. Nothing broke! :)

Are you saying that this is an optimization? Is that why it should be added back?

File: ocaml/typing/ctype.ml
Comment (goldfirere) [reply ocaml/typing/ctype.ml:2039]: Thanks about the `]`.

For the refactor: I don't think so, mostly because `Type_var` is neither success nor failure. The caller needs to figure out how to treat that.

File: ocaml/typing/mode_intf.mli
Comment (riaqn) [reply ocaml/typing/mode_intf.mli:29]: Ahh I see. Do you think it might be better to define inside `jkind.ml`:
```
val alloc_const_less_or_equal : ...
```
who is defined based on `Alloc.Const.le a b` and `Alloc.Const.le b a`?

Reasons:
- Adding `less_or_equal` to all axes adds lots of code, which becomes heavier as we add more axes. 
- `less_or_equal` is not a canonical interface of a lattice (`le` is), and used only by `jkind`.
- Is this `alloc_const_less_or_equal` slower? I think "two small pattern match vs. a single big pattern match" is hard to predict.
- Is `le` going through `less_or_equal` slower than a direct `le`? I thought inlining would make then same; but actually not - see my thread in the internal channel. That means the solver would be made slower.

File: ocaml/typing/jkind.mli
Comment (riaqn) [reply ocaml/typing/jkind.mli:40]: I agree. Let's just leave it here. Resolved.

File: ocaml/typing/mode_intf.mli
Comment (goldfirere) [reply ocaml/typing/mode_intf.mli:29]: I don't see why `less_or_equal` is any more or less canonical than `le`. Each can be implemented in terms of the other. Mathematicians are more likely to think about `le`, but I don't think that matters very much here.

The most performant thing will be to write all of these functions directly, not in terms of one another.

But I agree that this is not a big concern.

I think mostly this doesn't matter. And so that means the best thing is to stop debating. So I'll just do what you ask.

File: ocaml/testsuite/tests/typing-local/local.ml
Comment (goldfirere) [reply ocaml/testsuite/tests/typing-local/local.ml:1823]: Well, I thought this was OK, because we could say that we don't know the type of `y` principally. But I think that's a bit bogus. I will fix.

File: ocaml/typing/ctype.ml
Comment (goldfirere) [reply ocaml/typing/ctype.ml:5704]: I believe there is a principality check here.

File: ocaml/testsuite/tests/typing-modal-kinds/basics.ml
Comment (goldfirere) [reply ocaml/testsuite/tests/typing-modal-kinds/basics.ml:273]: I'm not sure I agree about the comment about the title (this really does use modal kinds to control mode-crossing), but yes I plan on fixing that stuff in future work.

File: ocaml/testsuite/tests/typing-modal-kinds/basics.ml
Comment (goldfirere) [reply ocaml/testsuite/tests/typing-modal-kinds/basics.ml:146]: Well, all the tests are about modal kinds -- which goes well beyond immediacy. If there's a CR, that means future work, yes.

File: ocaml/testsuite/tests/typing-local/local.ml
Comment (goldfirere) [reply ocaml/testsuite/tests/typing-local/local.ml:1823]: After looking into this for a while, I think my original behavior -- no mode-crossing in principal mode -- is correct. The reason is that `y`'s type is not known principally. We can observe this in this program:

```ocaml
type _ g =
  | Int : int g

let f (type a) (x : a g) (y : a) =
  match x, y with
  | Int, (3 as x) -> x
  | _ -> assert false
```

If the type of `x` were known principally, then we could infer the return type of `f` as `int`. But we currently get 

```
Error: This expression has type a = int
       but an expression was expected of type 'a 
       This instance of int is ambiguous:
       it would escape the scope of its equation
```

I know this because I fixed the bug, and the behavior here changed. So I favor just keeping the behavior as originally committed. If anything, we could treat the old behavior, allowing mode-crossing, as a bug.

Note that this does not affect code within Jane Street at all, because we do not use principal mode.

**Comment by goldfirere** [comment]
Ready for re-review here. (I still need to rebase, but I'll wait until review settles.)

File: ocaml/typing/ctype.ml
Comment (riaqn) [reply ocaml/typing/ctype.ml]: That makes sense.

File: ocaml/typing/ctype.ml
Comment (riaqn) [reply ocaml/typing/ctype.ml:5502]: Yes I think it's an optimization - in `typecore.ml`, `enlarge_type` is called first, and if failed, `subtype` is called. There is still mode crossing in `subtype`, so the lack of mode crossing in `enlarge_type` is not testable. 

I think it's doable but I don't know how to keep it tested in the long term.

For now, maybe just leave a CR here?

File: ocaml/typing/ctype.ml
Comment (riaqn) [reply ocaml/typing/ctype.ml:5704]: Do you mean there is no principality check here? I think you are right - the original PR(https://github.com/ocaml-flambda/flambda-backend/pull/1701) didn't check for principality for certain reason.

File: ocaml/typing/ctype.ml
Comment (lpw25) [reply ocaml/typing/ctype.ml:5502]: This isn't correct. `enlarge_type` is used when we are doing `(e :> t)` and we don't have (principally) know the type of `e`. Whereas `subtype` is used when we know the type of `e` (or if the user wrote `(e : t1 :> t2)`). I suspect something like this might be broken by this change:
```ocaml
module M : sig
  val foo : (local int -> unit) -> (int -> unit)
end = struct
  let foo f = (f :> int -> unit)
end
```

Comment (riaqn) [reply ocaml/typing/ctype.ml:5502]: thanks - I will add the missing logic with the test.

File: ocaml/testsuite/tests/typing-local/local.ml
Comment (riaqn) [reply ocaml/testsuite/tests/typing-local/local.ml:1823]: I agree that we should only cross mode if `is_principal ty = true`. But I'm slightly confused as to what's causing the change in output. 
Do you mean there is a BUG (independent of mode crossing) wrt principality check?

File: ocaml/typing/jkind.ml
Comment (riaqn) [reply ocaml/typing/jkind.ml]: For the monadic axes (e.g. uniqueness), I'm not sure what's the best way to characterise them. A value being `unique` means "you can strong update it if it's a record" or "it has no other aliases"? Consider the example: 
```
type t = { x : Int63.t} [@@unboxed}
```
This is immediate only on 64-bit systems. If we allow it cross to `unique` even on 32-bit system, then we should disallow strong update on this type.

(Similar for the `contention` axis).

I have an idea which is probably silly: can we make kinds platform-dependent? That is, we will only have a single `immediate` kind, and writing `type t : immediate = Int63.t` will error on 32bit systems?

Comment (goldfirere) [reply ocaml/typing/jkind.ml]: I implicitly assumed (but actually did think about) that we couldn't in-place update unboxed records. In-place update is about reusing memory... but an unboxed record doesn't have memory to reuse, so I'm not exactly sure what it would mean.

_Can_ we make kinds platform-dependent? Sure. _Should_ we make kinds platform-dependent? Almost certainly not. We don't want our type system to not be portable across machines. That's a recipe for devops sadness.

I agree there may be trouble in this area someday. But that day is not today!

File: ocaml/typing/ctype.ml
Comment (goldfirere) [reply ocaml/typing/ctype.ml]: Yes looks good. Thanks!

File: ocaml/testsuite/tests/typing-local/local.ml
Comment (goldfirere) [reply ocaml/testsuite/tests/typing-local/local.ml:1823]: There is a design choice that, in `match blah with 0 as x -> ...`, the type of `x` is not known principally. This makes some sense in the context of GADT pattern-matching, where principality means "the user has made their intent unambiguously clear".

Given this design choice -- and our desire to mode-cross only when a type is principally known -- we shouldn't mode-cross `x` either. Before this patch, we did mode-cross `x`, because there was a missing principality check. Now we don't. In light of the overall design here, I think this is an improvement.

**Comment by goldfirere** [comment]
@riaqn should review [these three commits](https://github.com/ocaml-flambda/flambda-backend/pull/2310/files/999dd5fccd2075861f814e332910ff5ebf632c55..acff688eb7650bee405ac9355f60deb774a1df40), and then we're otherwise good to go here.

File: ocaml/testsuite/tests/typing-local/local.ml
Comment (riaqn) [reply ocaml/testsuite/tests/typing-local/local.ml:1823]: Thanks for explanation.

File: ocaml/typing/jkind.ml
Comment (riaqn) [reply ocaml/typing/jkind.ml]: I'm interested in extending the conversation about platform-dependent kinds. But let's not block this PR.

**Review by riaqn — APPROVED** [review]
Thanks - I think it's good to go now.

**Pull request merged**
